### PR TITLE
Fix MRO to return 405 for unsupported methods

### LINKED ISSUE #2374: Prevent failing at /v1/auth_tokens/token/ due to `MethodNotAllowed`
## Description

The token endpoint fails with `500 Internal Server Error` when the method is different to POST. It should return a 405 error instead.

## Reproduction

1. Spin up the API `just api/up`
2. Try a request with any method different to POST, for example:
```sh
curl --request GET --url http://localhost:50280/v1/auth_tokens/token/
```
3. See error.

## Additional context

[Sentry context](https://openverse.sentry.io/share/issue/1f39e7a8ddb24891b449c25e009da88f/).

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #2374 by @krysal

## Description

This PR fixes the method resolution order of the token `APIView` so that the exception handler will be invoked and a 405 Method Not Allowed response will be sent. When the MRO is incorrect the exception is not handled and results in a 500 Internal Server Error.

## Testing Instructions

0. Check out the PR and run the dev server.
1. Make a GET request to the `/v1/auth_tokens/token/` endpoint.
2. Receive a 405 response.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/api/views/oauth2_views.py [hunk 1]
```diff
@@ -136,9 +137,9 @@ def get(self, request, code, format=None):
 
 
 @extend_schema(tags=["auth"])
-class TokenView(BaseTokenView, APIView):
+class TokenView(APIView, BaseTokenView):
     @token
-    def post(self, *args, **kwargs):
+    def post(self, request, *args, **kwargs):
```
Comment (krysal) [inline api/api/views/oauth2_views.py:142]: Looks like after `request` the arguments aren't needed anyomre.
```suggestion
    def post(self, request):
```

Comment (dhruvkb) [reply api/api/views/oauth2_views.py]: Sure, I thought it was best to keep the `*args` and `**kwargs` params when overridding methods but they are truly not needed in this case.

File: api/api/views/oauth2_views.py [hunk 2]
```diff
@@ -154,7 +155,9 @@ def post(self, *args, **kwargs):
         endpoint.
         """
 
-        return super().post(*args, **kwargs)
+        res = super().post(request._request)
+        data = json.loads(res.content)
+        return Response(data, status=res.status_code)
```
Comment (sarayourfriend) [inline api/api/views/oauth2_views.py:160]: Is it necessary to `json.loads` the content? Would something like this work without the need for that?

```suggestion
        data = json.loads(res.content)
        return HttpResponse(content=res.content, content_type="application/json", status=res.status_code)
```

`HttpResponse` is the more low-level version from `django.http` that allows you to bypass the DRF `Response` data parsing. Doing `json.loads` and then passing it directly to `Response` means we're unnecessary marshalling the JSON string to Python and then immediately back to a string.

Comment (sarayourfriend) [reply api/api/views/oauth2_views.py:160]: Although, more general question: why is it necessary to do anything manual here at all, other than just return `res` like we used to? Does `APIView` complain about not receiving the right type of response object or something?

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
Just one small request to remove some unnecessary (at least I _think_, could be wrong) JSON parsing. Krystle's change to the unit test is also critical too :slightly_smiling_face: 

Otherwise LGTM!

File: api/api/views/oauth2_views.py (hunk 2 above)
Comment (dhruvkb) [reply api/api/views/oauth2_views.py:160]: > Using the `Response` class simply provides a nicer interface for returning content-negotiated Web API responses, that can be rendered to multiple formats.
> &mdash; DRF docs

We can use `HttpResponse` for sure, but sending DRF's Response is more consistent as it renders the DRF API UI in the browser.

Comment (sarayourfriend) [reply api/api/views/oauth2_views.py:160]: > We can use HttpResponse for sure, but sending DRF's Response is more consistent and it renders the DRF API UI in the browser like our other endpoints.

Ahh, interesting, I didn't realise that was a side effect. For this view, it doesn't seem like a tragic loss, though? It's not exactly a "browseable" response :thinking: At least not any less so if your browser just renders a nice JSON explorer (Firefox does this, not sure of others).

I also don't think we _want_ to support anyone using this view over anything than JSON anyway, but I guess `Response` handles it regardless so :shrug: I would personally use HttpResponse but if you prefer to keep it, all good with me as well :+1:

**Review by sarayourfriend — APPROVED** [review]

**Review by krysal — APPROVED** [review]
Excellent! Thank you, @dhruvkb.

**Pull request merged**